### PR TITLE
Updating sequencer config implementation

### LINKED ISSUE #245: update configuration of astria-sequencer to config standard

### PR DESCRIPTION
**Pull request by dereksione**
## Summary
Updating how the sequencer does configuration management to use the specs defined in the configuration specs

## Background
Addressing the tech debt of different crates using different configuration setups and hierarchies

## Changes
- Removed manual env parsing in favor of figment
- Removed clap parsing
- Added justfile
- Added example .env
- Added common TESTNET commands to justfile
- Combined README and TESTNET file

## Testing
Added new figment tests for the config setup

## Breaking Changelist
- All config management should be done using .env files now, as outlined in the spec - cli overrides support is removed

## Related Issues
Link any issues that are related, prefer full github links.

closes #245

### CODE REVIEW

File: crates/astria-sequencer/src/main.rs [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 
 #[tokio::main]
 async fn main() {
-    let config = Config::get();
-    telemetry::init(std::io::stdout, config.log.as_deref().unwrap_or("info"))
-        .expect("failed to initialize telemetry");
+    let config = match config::get() {
+        Ok(cfg) => cfg,
+        Err(e) => {
+            eprintln!("failed to read configuration: {e}");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/main.rs:12]: Print the entire error trace with `{e:?}`

**Review by SuperFluffy — CHANGES_REQUESTED** [review]
Quick review for some issues I saw. It's great we are removing a lot of code

File: crates/astria-sequencer/justfile [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  tail -f sequencer_testnet/{{ NODE }}/cometbft.log
+
+sequencer-logs-testnet NODE="node0":
+  tail -f sequencer_testnet/{{ NODE }}/sequencer.log
+
+stop-testnet:
+  pkill cometbft && pkill astria-sequencer
+
+clean-testnet DIR="sequencer_testnet": stop-testnet
+  rm -r {{ DIR }}
```
Comment (joroshiba) [inline crates/astria-sequencer/justfile:44]: nit: new line

File: crates/astria-sequencer/README.md [hunk 3]
```diff
@@ -1,57 +1,71 @@
...
 
 Because `penumbra-storage:0.54.1` depends on `rocksdb:0.19.0` compilation on gcc
 13 will not work.
 
 On arch linux:
 
-```sh
-sudo pacman -S gcc12 gcc12-libs
-CC=/usr/bin/gcc-12 CXX=/usr/bin/c++-12 cargo build
+```bash
```
Comment (noot) [inline crates/astria-sequencer/README.md:19]:
```suggestion
```sh
```

same with others in this file for consistency

File: crates/astria-sequencer/README.md [hunk 4]
```diff
@@ -69,10 +83,66 @@ I[2023-05-16|16:53:56.786] service start    module=abci-client
 ```sh
 # initialize the node
 cometbft init
-# set the block time to 15s
-sed -i'.bak' 's/timeout_commit = "1s"/timeout_commit = "15s"/g' ~/.cometbft/config/config.toml
+# set the block time to 14s
```
Comment (noot) [inline crates/astria-sequencer/README.md:86]:
```suggestion
# set the block time to 15s
```

File: crates/astria-sequencer/README.md [hunk 5]
```diff
@@ -69,10 +83,66 @@ I[2023-05-16|16:53:56.786] service start    module=abci-client
...
+sed -i'.bak' 's/timeout_commit = "0s"/timeout_commit = "15s"/g' ~/.cometbft/config/config.toml
```
Comment (noot) [inline crates/astria-sequencer/README.md:87]: why is this changed?

File: crates/astria-sequencer/README.md [hunk 6]
```diff
@@ -69,10 +83,66 @@ I[2023-05-16|16:53:56.786] service start    module=abci-client
...
 
 You should see blocks being produced.
+
+You can also use `just` to run the above commands:
+
+```bash
+just run-cometbft
+```
+
+## Testnet
```
Comment (noot) [inline crates/astria-sequencer/README.md:100]: I kinda liked having the testnet instructions in a different file, maybe just add a link to it here instead?

File: crates/astria-sequencer/local.env.example [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+ASTRIA_COMPOSER_LOG="astria_composer=info"
+
+# Address of the API server
+ASTRIA_SEQUENCER_GENESIS_FILE="test-genesis.json"
+
+# Address of the RPC server for the sequencer chain
+ASTRIA_SEQUENCER_LISTEN_ADDR="http://127.0.0.1:26658"
+
+# Path to rocksdb
+ASTRIA_SEQUENCER_DB_FILEPATH="../"
```
Comment (noot) [inline crates/astria-sequencer/local.env.example:11]: if this is supposed to be a directory (which I think it should be) rename this to `ASTRIA_SEQUENCER_DB_DATADIR`?

File: crates/astria-sequencer/local.env.example [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Log level for the sequencer
+ASTRIA_COMPOSER_LOG="astria_composer=info"
+
+# Address of the API server
+ASTRIA_SEQUENCER_GENESIS_FILE="test-genesis.json"
+
+# Address of the RPC server for the sequencer chain
```
Comment (noot) [inline crates/astria-sequencer/local.env.example:7]:
```suggestion
# Address of the RPC server for the sequencer ABCI application
```

File: crates/astria-sequencer/local.env.example [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Log level for the sequencer
+ASTRIA_COMPOSER_LOG="astria_composer=info"
+
+# Address of the API server
```
Comment (noot) [inline crates/astria-sequencer/local.env.example:4]: update comment

File: crates/astria-sequencer/src/config.rs
Comment (SuperFluffy) [inline crates/astria-sequencer/src/config.rs:1]: Let's update this to match the changes introduced with https://github.com/astriaorg/astria/pull/387

File: crates/astria-sequencer/src/config.rs [hunk 10]
```diff
@@ -1,40 +1,71 @@
...
     /// The endpoint on which Sequencer will listen for ABCI requests
-    #[arg(long, default_value_t = String::from(DEFAULT_ABCI_LISTEN_ADDR))]
     pub listen_addr: String,
     /// The path to penumbra storage db.
-    #[arg(long)]
-    pub db_filepath: PathBuf,
-    /// Filter directives for emitting events
-    #[arg(long)]
-    pub log: Option<String>,
+    pub data_dir: PathBuf,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/config.rs:23]: MO this PR shouldn't change the name of this option. Reading the penumbra-storage code [1] and the rocksdb [2] docs it also does not expect a directory, but a path to a file.

1: https://rustdoc.penumbra.zone/main/src/penumbra_storage/storage.rs.html#46
2: https://docs.rs/rocksdb/0.21.0/rocksdb/type.DBWithThreadMode.html#method.open_cf

File: crates/astria-sequencer/src/sequencer.rs [hunk 11]
```diff
@@ -26,21 +26,21 @@ impl Sequencer {
...
             );
         } else {
             info!(
-                path = %config.db_filepath.display(),
+                path = %config.data_dir.display(),
                 "creating storage db"
             );
         }
-        let storage = penumbra_storage::Storage::load(config.db_filepath.clone())
+        let storage = penumbra_storage::Storage::load(config.data_dir.clone())
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/sequencer.rs:43]: See above, `data_dir` is probably incorrectly named.

File: crates/astria-sequencer/src/main.rs [hunk 12]
```diff
@@ -1,17 +1,26 @@
...
 #[tokio::main]
 async fn main() {
-    let config = Config::get();
-    telemetry::init(std::io::stdout, config.log.as_deref().unwrap_or("info"))
-        .expect("failed to initialize telemetry");
+    let config = match config::get() {
+        Ok(cfg) => cfg,
+        Err(e) => {
+            eprintln!("failed to read configuration: {e:?}");
+            std::process::exit(EX_CONFIG);
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/main.rs:17]: I have had a read about `process::exit` and think we should avoid it, because it immediately stops the process without running any destructors. That's not a huge deal in this case, but seems bad practice in general (future PRs might add some code before the config result gets evaluated, not paying attention the exit).

It's probably better to change `async fn main() -> std::process::ExitCode`

Have a look at the conductor refactor here and what its main.rs does: https://github.com/astriaorg/astria/pull/392

File: crates/astria-sequencer/src/main.rs [hunk 13]
```diff
@@ -1,17 +1,26 @@
...
+
 #[tokio::main]
 async fn main() {
-    let config = Config::get();
-    telemetry::init(std::io::stdout, config.log.as_deref().unwrap_or("info"))
-        .expect("failed to initialize telemetry");
+    let config = match config::get() {
+        Ok(cfg) => cfg,
+        Err(e) => {
+            eprintln!("failed to read configuration: {e:?}");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/main.rs:16]:
```suggestion
            eprintln!("failed to read configuration:\n{e:?}");
```

File: crates/astria-sequencer/local.env.example [hunk 14]
```diff
@@ -0,0 +1,8 @@
+# Log level for the sequencer
+ASTRIA_SEQUENCER_LOG="astria_composer=info"
+
+# Address of the RPC server for the sequencer ABCI application
```
Comment (SuperFluffy) [inline crates/astria-sequencer/local.env.example:4]:
```suggestion
# Socket address to listen for ABCI requests from cometbft.
# This address corresponds to the `--proxy_app "tcp://<ASTRIA_SEQUENCER_LISTEN_ADDR>"`,
# where `tcp://127.0.0.1:26658` is comebft's default.
```

File: crates/astria-sequencer/local.env.example [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+ASTRIA_SEQUENCER_LISTEN_ADDR="http://127.0.0.1:26658"
```
Comment (SuperFluffy) [inline crates/astria-sequencer/local.env.example:5]: This expects somehing that can be parsted as a `std::net::SocketAddr` without the protocol.
```suggestion
ASTRIA_SEQUENCER_LISTEN_ADDR="127.0.0.1:26658"
```

File: crates/astria-sequencer/local.env.example [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+
+# Path to rocksdb
+ASTRIA_SEQUENCER_DATA_DIR="../"
```
Comment (SuperFluffy) [inline crates/astria-sequencer/local.env.example:8]: This is a misnomer: the code expects a file path, not a dir.


(also nit: newline)

File: crates/astria-sequencer/README.md [hunk 17]
```diff
@@ -21,18 +24,32 @@ CC=/usr/bin/gcc-12 CXX=/usr/bin/c++-12 cargo build
...
 
 Ensure `~/go` is in your `PATH`, or `GOPATH` is set to some other place in your
 `PATH`.
 
 ```sh
-git clone https://github.com/astriaorg/cometbft
-cd cometbft
-export GOPATH=~/go
-make install
+just install-cometbft
```
Comment (SuperFluffy) [inline crates/astria-sequencer/README.md:52]: There is no such just command. Also, I would probably leave the steps in place. We don't want to mess with people's setup (e.g. if they have a different GOPATH, like myself).

File: crates/astria-sequencer/justfile [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+copy-env type=default_env:
+  cp {{ type }}.env.example .env
+
+run:
+  cargo run
+
+
+run-cometbft:
+  cometbft init
+  sed -i'.bak' 's/timeout_commit = "0s"/timeout_commit = "15s"/g' ~/.cometbft/config/config.toml
```
Comment (SuperFluffy) [inline crates/astria-sequencer/justfile:17]: I am a bit uncomfortable changing people's local files. What if they have a cometbft already present?

File: crates/astria-sequencer/local.env.example [hunk 19]
```diff
@@ -0,0 +1,8 @@
+# Log level for the sequencer
+ASTRIA_SEQUENCER_LOG="astria_composer=info"
```
Comment (SuperFluffy) [inline crates/astria-sequencer/local.env.example:2]:
```
ASTRIA_SEQUENCER_LOG="astria_sequencer=info"
```

**Review by SuperFluffy — CHANGES_REQUESTED** [review]
blocking because local.env.example is currently broken.

otherwise just a few minor comments

File: crates/astria-sequencer/justfile (hunk 18 above)
Comment (dereksione) [reply crates/astria-sequencer/justfile:17]: Would be open to alternate scripts - this is just what I rebased from main

Comment (SuperFluffy) [reply crates/astria-sequencer/justfile:17]: Yeah, good point. Setting up and taking down the env is certainly more complex than a just script allows. Maybe add a disclaimer?

File: crates/astria-sequencer/README.md [hunk 20]
```diff
@@ -21,15 +24,31 @@ CC=/usr/bin/gcc-12 CXX=/usr/bin/c++-12 cargo build
...
+just copy-env
+```
 
 ### Install cometbft
 
 Ensure `~/go` is in your `PATH`, or `GOPATH` is set to some other place in your
 `PATH`.
 
 ```sh
-git clone https://github.com/astriaorg/cometbft
```
Comment (SuperFluffy) [inline crates/astria-sequencer/README.md:32]: Don't rm this lin

File: crates/astria-sequencer/src/main.rs (hunk 13 above)
Comment (SuperFluffy) [reply crates/astria-sequencer/src/main.rs:18]: The newline would make this nicer to read..

**Review by SuperFluffy — APPROVED** [review]
Looks good. Ack with the justfile and setup script for cometbft. Should cover this in a future PR.

Left a note about installing cometbft in the README.

File: crates/astria-sequencer/src/main.rs (hunk 13 above)
Comment (dereksione) [reply crates/astria-sequencer/src/main.rs:18]: Oh sorry mb - oversight

File: crates/astria-sequencer/justfile [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+default_env := 'local'
+copy-env type=default_env:
+  cp {{ type }}.env.example .env
+
+run:
+  cargo run
+
+
+run-cometbft:
+  cometbft init
```
Comment (noot) [inline crates/astria-sequencer/justfile:16]: add this line after
```
../../target/debug/astria-sequencer-utils --genesis-app-state-file=test-genesis-app-state.json --destination-genesis-file=$HOME/.cometbft/config/genesis.json
```

File: crates/astria-sequencer/README.md [hunk 22]
```diff
@@ -69,12 +90,25 @@ I[2023-05-16|16:53:56.786] service start    module=abci-client
...
+
+You can also use `just` to run the above commands:
+
+```sh
+just run-cometbft
+```
+
+## Testnet
+
+Check out the `TESTNET.md` file for details on how to run the astria testnet
```
Comment (noot) [inline crates/astria-sequencer/README.md:114]:
```suggestion
Check out the `TESTNET.md` file for details on how to run a multi-node sequencer testnet.
```

File: crates/astria-sequencer/TESTNET.md [hunk 23]
```diff
@@ -28,37 +28,51 @@ This will create 3 cometbft validator configs and put the validator configs in
 `./sequencer_testnet` by default. It will then start the respective sequencer
 applications (one for each cometbft validator) and the cometbft validator nodes.
 
+(This requires cometbft to be installed)
+
 ```sh
-./scripts/testnet.sh
+just run-testnet
```
Comment (noot) [inline crates/astria-sequencer/TESTNET.md:34]: i liked the instructions better before, this is gonna rebuild every time (even though it should be fast but still)

File: crates/astria-sequencer/TESTNET.md [hunk 24]
```diff
@@ -28,37 +28,51 @@ This will create 3 cometbft validator configs and put the validator configs in
...
 ## Cleaning up the testnet
 
-Simply stop the testnet and `rm -r sequencer_testnet` or whichever directory the
-config files are in.
+Simply stop the testnet and run
+
+```sh
+just clean-testnet sequencer_testnet
+```
+where `sequencer_testnet` is the dir where the config exists
```
Comment (noot) [inline crates/astria-sequencer/TESTNET.md:78]:
```suggestion
where `sequencer_testnet` is the dir where the testnet files are.
```

File: crates/astria-sequencer/TESTNET.md [hunk 25]
```diff
@@ -28,37 +28,51 @@ This will create 3 cometbft validator configs and put the validator configs in
...
 ```sh
-./scripts/testnet.sh
+just run-testnet
 ```
 
 To manually set the output directory or number of validators:
 
 ```sh
-OUT_DIR=out_dir NUM_VALIDATORS=num_validators ./scripts/testnet.sh
+OUT_DIR=out_dir NUM_VALIDATORS=num_validators just run-testnet
```
Comment (noot) [inline crates/astria-sequencer/TESTNET.md:40]: same for here

File: crates/astria-sequencer/local.env.example [hunk 26]
```diff
@@ -0,0 +1,10 @@
+# Log level for the sequencer
+ASTRIA_SEQUENCER_LOG="astria_sequencer=info"
+
+# Socket address to listen for ABCI requests from cometbft.
+# This address corresponds to the `--proxy_app "tcp://<ASTRIA_SEQUENCER_LISTEN_ADDR>"`,
+# where `tcp://127.0.0.1:26658` is comebft's default.
+ASTRIA_SEQUENCER_LISTEN_ADDR="127.0.0.1:26658"
+
+# Path to rocksdb
+ASTRIA_SEQUENCER_DB_FILEPATH="../db"
```
Comment (noot) [inline crates/astria-sequencer/local.env.example:10]:
```suggestion
ASTRIA_SEQUENCER_DB_FILEPATH="/tmp/astria_db"
```

otherwise it'll go in `crates` which seems bad

File: crates/astria-sequencer/src/config.rs [hunk 27]
```diff
@@ -1,40 +1,91 @@
...
+};
 use serde::{
     Deserialize,
     Serialize,
 };
 
-const DEFAULT_ABCI_LISTEN_ADDR: &str = "127.0.0.1:26658";
+/// # Errors:
+///
+/// returns an error generated by figment highlighting what went wrong in trying to read the config
```
Comment (noot) [inline crates/astria-sequencer/src/config.rs:14]:
```suggestion
/// - if figment failed to read the config from the environment
```

File: crates/astria-sequencer/TESTNET.md (hunk 23 above)
Comment (dereksione) [reply crates/astria-sequencer/TESTNET.md:34]: Cargo doesn't rebuild when things are cached and there are no changes, no?

File: crates/astria-sequencer/README.md [hunk 28]
```diff
@@ -52,6 +72,7 @@ cargo build
 ../../target/debug/astria-sequencer --db-filepath=/tmp/astria_db
```
[the remark this answers is not in the document]
Comment (dereksione) [reply crates/astria-sequencer/README.md:72]: Good catch

**Review by noot — APPROVED** [review]
